File: backend/models/authorization.py
"""人员授权记录。"""
from datetime import datetime

from . import db


class Authorization(db.Model):
    __tablename__ = "authorizations"

    id = db.Column(db.Integer, primary_key=True)
    grantee_username = db.Column(db.String(100), nullable=False, index=True)
    grantee_name = db.Column(db.String(50), nullable=False, default="")
    grantee_dept_code = db.Column(db.String(20), nullable=False, default="", index=True)
    source = db.Column(db.String(20), nullable=False)
    granter_name = db.Column(db.String(50), nullable=False, default="")
    granter_dept_code = db.Column(db.String(20), nullable=False, default="", index=True)
    # 委托是否继续有效必须和授权当时的负责人比，不能只保存当前负责人姓名。
    granter_head_snapshot = db.Column(db.String(50), nullable=False, default="")
    doc_no = db.Column(db.String(100), nullable=False, default="")
    perm_keys = db.Column(db.Text, nullable=False, default="[]")
    valid_from = db.Column(db.String(10), nullable=False)
    valid_to = db.Column(db.String(10), nullable=False)
    doc_path = db.Column(db.Text, nullable=False)
    doc_name = db.Column(db.String(255), nullable=False)
    status = db.Column(db.String(20), nullable=False, default="active", index=True)
    created_by = db.Column(db.String(100), nullable=False, default="")
    created_at = db.Column(db.DateTime, nullable=False, default=datetime.now)
    revoked_by = db.Column(db.String(100), nullable=False, default="")
    revoked_at = db.Column(db.DateTime, nullable=True)
    revoke_reason = db.Column(db.Text, nullable=False, default="")

File: frontend/src/services/authorization.ts
import api from './api'

export interface PermItem { key: string; label: string }
export interface PermGroup { group: string; items: PermItem[] }
export interface AuthzUser {
  username: string; display_name: string; dept_code: string; role: string; active: number
}
export interface AuthzDept { code: string; name: string; head_name: string; active: number }
export interface AuthorizationInfo {
  id: number
  grantee_username: string
  grantee_name: string
  grantee_dept_code: string
  source: 'resolution' | 'delegate'
  granter_name: string
  granter_dept_code: string
  doc_no: string
  perm_keys: string[]
  valid_from: string
  valid_to: string
  doc_name: string
  status: string
  effective_state: string
  created_at: string
  revoked_by: string
  revoked_at: string
  revoke_reason: string
}

export const listAuthorizations = (params?: Record<string, string>) =>
  api.get<{ ok: boolean; data: AuthorizationInfo[] }>('/authorizations', { params })
export const myAuthorizations = () =>
  api.get<{ ok: boolean; data: AuthorizationInfo[] }>('/authorizations/my')
export const getAuthzCatalog = () =>
  api.get<{ ok: boolean; data: PermGroup[] }>('/authorizations/perm-catalog')
export const getAuthzUsers = () =>
  api.get<{ ok: boolean; data: AuthzUser[] }>('/authorizations/users')
export const getAuthzDepts = () =>
  api.get<{ ok: boolean; data: AuthzDept[] }>('/authorizations/depts')
export const uploadAuthzDocument = (file: File) => {
  const body = new FormData()
  body.append('file', file)
  return api.post<{ ok: boolean; path: string; name: string }>('/authorizations/upload', body)
}
export const createAuthorization = (data: Record<string, unknown>) =>
  api.post<{ ok: boolean; data: AuthorizationInfo }>('/authorizations', data)
export const revokeAuthorization = (id: number, reason: string) =>
  api.post<{ ok: boolean }>(`/authorizations/${id}/revoke`, { reason })
export const authzDocumentUrl = (id: number) => `/api/authorizations/${id}/document`

